File: base/algorithm/11S5Alg/DS11S5LotteryNumAlg.m
//
//  DS11S5LotteryNumAlg.m
//  DSBet
//
//  Created by Selena on 2018/1/19.
//  Copyright © 2018年 Bill. All rights reserved.
//

#import "DS11S5LotteryNumAlg.h"

@implementation DS11S5LotteryNumAlg

//获取注数
+ (NSInteger)getBetNoteFromBetNum:(NSString *)betNumStr lotteryID:(NSInteger)lotteryID
{
    NSInteger betNum = 0;
    if(betNumStr.length ==0 ){
        return BET_NUM_ERROE;
    }
    
    switch (lotteryID) {
        case 2001: //前三直选复式
        {
            betNum = [self getBetNumStrZhixuan:betNumStr num:3];
            break;
            
            break;
        }
        case 2003: //前三组选复式      //C(n,3)
        case 2013:  //三中三
        {
            NSInteger num = [self getNumFromRow:0 withStr:betNumStr];
            betNum = [self getBetNumCTotalNum:num withNum:3];
            
            break;
            
            break;
        }
        case 2005:  //前二直选复式
        {
            betNum = [self getBetNumStrZhixuan:betNumStr num:2];
            break;
        }
        case 2007:  //前二组选复式   //C(n,2)
        case 2012:  //复式二中二
        {
            NSInteger num = [self getNumFromRow:0 withStr:betNumStr];
            betNum = [self getBetNumCTotalNum:num withNum:2];
            break;
        }
        case 2009:  //不位胆
        case 2011:  //复式一中一
        {
            NSInteger num = [self getNumFromRow:0 withStr:betNumStr];
            betNum = num;
            break;
        }
        case 2010:  //定位胆
        {
            NSString    *replacedStr = [betNumStr stringByReplacingOccurrencesOfString:COMMA_UTIL withString:SPACE_UTIL];
            NSArray *list = [replacedStr componentsSeparatedByString:SPACE_UTIL];
            NSInteger count = 0;
            for(NSInteger i = 0; i < list.count; i++){
                NSString *str = [list objectAtIndex:i];
                if(![str isEqualToString:GANG_UTIL]){
                    count++;
                }
            }
            betNum = count;
            break;
        }
        case 2014:
        {
            NSInteger num = [self getNumFromRow:0 withStr:betNumStr];
            betNum = [self getBetNumCTotalNum:num withNum:4];
            break;
        }
        case 2015:
        {
            NSInteger num = [self getNumFromRow:0 withStr:betNumStr];
            betNum = [self getBetNumCTotalNum:num withNum:5];
            break;
        }
        case 2016:
        {
            NSInteger num = [self getNumFromRow:0 withStr:betNumStr];
            betNum = [self getBetNumCTotalNum:num withNum:6];
            break;
        }
        case 2017:
        {
            NSInteger num = [self getNumFromRow:0 withStr:betNumStr];
            betNum = [self getBetNumCTotalNum:num withNum:7];
            break;
        }
        case 2018:
        {
            NSInteger num = [self getNumFromRow:0 withStr:betNumStr];
            betNum = [self getBetNumCTotalNum:num withNum:8];
            break;
        }
   
        default:
            break;
    }

    return betNum;
}

//获取注数  输入
+ (NSInteger)getPutInBetNoteFromBetNum:(NSString *)betNumStr lotteryID:(NSInteger)lotteryID needNum:(NSInteger)needNum
{
    NSInteger betNum = 0;
    if(betNumStr.length ==0 ){
        return BET_NUM_ERROE;
    }

    betNum = [self getBetNumPutIn:betNumStr needNum:needNum];
    NSLog(@"注数为:%ld",(long)betNum);
    return betNum;
    
    return betNum;
}

//获取正确的投注  最后的处理  输入
+ (NSString *)getPutCorrectBetNumStr:(NSString *)betNumStr lotteryID:(NSInteger)lotteryID ticketID:(NSInteger)ticketID needNum:(NSInteger)needNum
{
    NSString *betStr = betNumStr;
    
    betStr = [betStr stringByReplacingOccurrencesOfString:@"," withString:@""];
    betStr = [betStr stringByReplacingOccurrencesOfString:@";" withString:@""];
    betStr = [betStr stringByReplacingOccurrencesOfString:@" " withString:@""];
    NSMutableString *muStr = [NSMutableString new];
    if(betStr.length > 0){
        if(betStr.length%2 != 0){
            return betStr;
        }else{
            NSInteger length = betStr.length;
            for(NSInteger i = 0; i < (length/2); i ++){
                NSString *str = [betStr substringWithRange:NSMakeRange(i * 2, 2)];
                [muStr appendString:str];
                //TODO
                if(i != ((length/2) - 1)){
                    if(((i + 1)*2)%needNum == 0){
                        [muStr appendString:@";"];
                    }else{
                        [muStr appendString:@" "];
                    }
                    
                }
            }
            return muStr;
        }
    }

    return betStr;
}


#pragma mark  - 工具
#pragma mark - 过滤字符串
+ (NSString *)get11S5CorrectStrFromStr:(NSString *)numStr lotteryID:(NSInteger)lotteryID
{
    NSString *betNumStr = numStr;
    if([betNumStr rangeOfString:COMMA_UTIL].location !=NSNotFound)
    {
        NSMutableString *reStr = [[NSMutableString alloc] init] ;
        NSArray *list = [betNumStr componentsSeparatedByString:COMMA_UTIL];
        for(NSInteger i = 0; i < (list.count - 1); i ++){
            NSString *str = [list objectAtIndex:i];
            BOOL isSame = NO;
            for(NSInteger j = (i + 1);j < list.count;j ++){
                NSString *str2 = [list objectAtIndex:j];
                if([str isEqualToString:str2]){
                    isSame = YES;
                    break;
                }else if([self isBeyond:str])
                {
                    isSame = YES;
                    break;
                }
            }
            if((!isSame) && ([self isPassPaozi:str playId:lotteryID]))
            {
                [reStr appendString:str];
                [reStr appendString:@","];
            }
        }
        NSString *lastStr = [list lastObject];
        
        if((![self isBeyond:lastStr]) && ([self isPassPaozi:lastStr playId:lotteryID])){
            [reStr appendString:lastStr];
            betNumStr = reStr;
        }else{
            betNumStr = reStr;
        }
        NSLog(@"11组选过滤前%@",betNumStr);
        betNumStr = [self isPassZuXuan:betNumStr playId:lotteryID];
    }else{
        if([self isBeyond:betNumStr]){
            return @"";
        }
    }
    return betNumStr;
}

+ (BOOL)isBeyond:(NSString *)numStr
{
    BOOL isBeyond = NO;
    NSLog(@"过滤");
    if(numStr.length > 0){
        if(numStr.length%2 != 0){
            return YES;
        }else{
            NSInteger length = numStr.length;
            for(NSInteger i = 0; i < (length/2); i ++){
                NSString *str = [numStr substringWithRange:NSMakeRange(i * 2, 2)];
                NSInteger num = [str integerValue];
                if((num <= 0) || (num >= 12)){
                    isBeyond = YES;
                    break;
                }
            }
        }
    }
    return isBeyond;
}

#pragma mark - 获取正确的投注
+ (NSString *)getCorrectBetNumStr:(NSString *)betNumStr playID:(long)playId
{
    NSString *betStr = betNumStr;
    if([betNumStr rangeOfString:COMMA_UTIL].location !=NSNotFound)
    {
        NSArray *list = [betNumStr componentsSeparatedByString:COMMA_UTIL];
        NSMutableString *numStrNum = [NSMutableString new];
        for(NSString *str in list){
            NSInteger length = str.length;
            if(length%2 != 0){
                return nil;
                break;
            }
            
            for(NSInteger i = 0; i < (length/2); i ++){
                NSString *str2 = [str substringWithRange:NSMakeRange(i * 2, 2)];
                if(i != (length/2 - 1)){
                    str2 = [NSString stringWithFormat:@"%@ ",str2];
                }
                [numStrNum appendString:str2];
            }
            [numStrNum appendString:@","];
        }
        betStr = [numStrNum substringToIndex:(numStrNum.length - 1)];
        
    }else{
        NSMutableString *numStrNum = [NSMutableString new];
        for(NSInteger i = 0; i < (betNumStr.length/2); i ++){
            NSString *str2 = [betNumStr substringWithRange:NSMakeRange(i * 2, 2)];
            if(i != (betNumStr.length/2 - 1)){
                str2 = [NSString stringWithFormat:@"%@ ",str2];
            }
            [numStrNum appendString:str2];
        }
        betStr = numStrNum;
    }
    return betStr;
}
+ (NSString *)isPassZuXuan:(NSString *)betStr playId:(long)playId
{
    switch (playId) {
            //单式
        case 2002:
        case 2006:
        case 2019:
        case 2020:
        case 2021:
        case 2022:
        case 2023:
        case 2024:
        case 2025:
        case 2026:

        {
            return betStr;
            break;
        }
            //组选   任选
            
        default:
        {
            NSString *reStr = [self isZuXuanNum:betStr isDouble:YES];
            return reStr;
            break;
        }
    }
    return @"";
}

+ (BOOL)isPassPaozi:(NSString *)betStr playId:(long)playId
{
    BOOL isAllSame = [self isAllSameNum2:betStr];
    if(isAllSame){
        return NO;
    }
    return YES;
}


@end

File: base/src/main/java/com/desheng/base/algorithm/d11s5/DS11S5LotteryNumAlg.java
package com.desheng.base.algorithm.d11s5;//
//  DS11S5LotteryNumAlg.m
//  DSBet
//
//  Created by Selena on 2018/1/19.
//  Copyright © 2018年 Bill. All rights reserved.
//

import com.ab.debug.AbDebug;
import com.ab.util.ArraysAndLists;
import com.ab.util.Strs;
import com.desheng.base.algorithm.DS11S5AlgorithObject;
import com.desheng.base.context.CtxLottery;

public class DS11S5LotteryNumAlg extends DS11S5AlgorithObject {
    
    /**
     * 获取注数
     */
    public static long getBetNoteFromBetNum(String
                                                    betNumStr, int lotteryID)
    
    {
        long betNum = 0;
        if (betNumStr.length() == 0) {
            return CtxLottery.BET_NUM_ERROE;
        }
        
        switch (lotteryID) {
            case 2001: //前三直选复式
            {
                betNum = getBetNumStrZhixuan(betNumStr, 3);
                break;
                
            }
            case 2003: //前三组选复式      //C(n,3)
            case 2013:  //三中三
            {
                int num = getNumFromRow(0, betNumStr);
                betNum = getBetNumCTotalNum(num, 3);
                
                break;
                
            }
            case 2005:  //前二直选复式
            {
                betNum = getBetNumStrZhixuan(betNumStr, 2);
                break;
            }
            case 2007:  //前二组选复式   //C(n,2)
            case 2012:  //复式二中二
            {
                int num = getNumFromRow(0, betNumStr);
                betNum = getBetNumCTotalNum(num, 2);
                break;
            }
            case 2009:  //不位胆
            case 2011:  //复式一中一
            {
                int num = getNumFromRow(0, betNumStr);
                betNum = num;
                break;
            }
            case 2010:  //定位胆
            {
                String replacedStr = betNumStr.replace(",", " ");
                String[] list = replacedStr.split(" ");
                long count = 0;
                for (int i = 0; i < list.length; i++) {
                    String str = list[i];
                    if (!str.equals("-")) {
                        count++;
                    }
                }
                betNum = count;
                break;
            }
            case 2014: {
                int num = getNumFromRow(0, betNumStr);
                betNum = getBetNumCTotalNum(num, 4);
                break;
            }
            case 2015: {
                int num = getNumFromRow(0, betNumStr);
                betNum = getBetNumCTotalNum(num, 5);
                break;
            }
            case 2016: {
                int num = getNumFromRow(0, betNumStr);
                betNum = getBetNumCTotalNum(num, 6);
                break;
            }
            case 2017: {
                int num = getNumFromRow(0, betNumStr);
                betNum = getBetNumCTotalNum(num, 7);
                break;
            }
            case 2018: {
                int num = getNumFromRow(0, betNumStr);
                betNum = getBetNumCTotalNum(num, 8);
                break;
            }
            
            default:
                break;
        }
        
        return betNum;
    }
    
    //获取注数  输入
    public static long getPutInBetNoteFromBetNum(String betNumStr, int lotteryID, int needNum)
    
    {
        long betNum = 0;
        if (betNumStr.length() == 0) {
            return CtxLottery.BET_NUM_ERROE;
        }
        
        betNum = getBetNumPutIn(betNumStr, needNum);
        AbDebug.log(AbDebug.TAG_APP, String.format("注数为:%d", (long) betNum));
        return betNum;
    }
    
    //获取正确的投注  最后的处理  输入
    public static String getPutCorrectBetNumStr(String betNumStr, int lotteryID, int ticketID, int needNum)
    
    {
        String betStr = betNumStr;
        
        betStr = betStr.replace(",", "");
        betStr = betStr.replace(";", "");
        betStr = betStr.replace(" ", "");
        StringBuffer muStr = new StringBuffer();
        if (betStr.length() > 0) {
            if (betStr.length() % 2 != 0) {
                return betStr;
            } else {
                long length = betStr.length();
                for (int i = 0; i < (length / 2); i++) {
                    String str = betStr.substring(i * 2, i * 2 + 2);
                    muStr.append(str);
                    if (i != ((length / 2) - 1)) {
                        if (((i + 1)*2)%needNum == 0) {
                            muStr.append(";");
                        } else {
                            muStr.append(" ");
                        }
                    }
                }
                return muStr.toString();
            }
        }
        
        return betStr;
    }
    
    
    /**
     * -过滤字符串
     */
    public static String get11S5CorrectStrFromStr(String numStr, int lotteryID)
    
    {
        String betNumStr = numStr;
        if (betNumStr.contains(",")) {
            StringBuffer reStr = new StringBuffer();
            String[] list = betNumStr.split(",");
            for (int i = 0; i < list.length - 1; i++) {
                String str = list[i];
                boolean isSame = false;
                for (int j = (i + 1); j < list.length; j++) {
                    String str2 = list[j];
                    if (str.equals(str2)) {
                        isSame = true;
                        break;
                    } else if (isBeyond(str)) {
                        isSame = true;
                        break;
                    }
                }
                if ((!isSame) && (isPassPaozi(str, lotteryID))) {
                    reStr.append(str);
                    reStr.append(",");
                }
            }
            String lastStr = ArraysAndLists.lastObject(list);
            
            if (!isBeyond(lastStr) && (isPassPaozi(lastStr, lotteryID))) {
                reStr.append(lastStr);
                betNumStr = reStr.toString();
            } else {
                betNumStr = reStr.toString();
            }
            AbDebug.log(AbDebug.TAG_APP, String.format("11组选过滤前%s", betNumStr));
            betNumStr = isPassZuXuan(betNumStr, lotteryID);
        } else {
            if (isBeyond(betNumStr)) {
                return "";
            }
        }
        return betNumStr;
    }
    
    public static boolean isBeyond(String numStr)
    
    {
        boolean isBeyond = false;
        AbDebug.log(AbDebug.TAG_APP, String.format("过滤"));
        if (numStr.length() > 0) {
            if (numStr.length() % 2 != 0) {
                return true;
            } else {
                long length = numStr.length();
                for (int i = 0; i < (length / 2); i++) {
                    String str = numStr.substring(i * 2, i * 2 + 2);
                    int num = Strs.parse(str, -1);
                    if ((num <= 0) || (num >= 12)) {
                        isBeyond = true;
                        break;
                    }
                }
            }
        }
        return isBeyond;
    }
    
    /**
     * pragma mark -获取正确的投注
     */
    public static String getCorrectBetNumStr(String betNumStr, int playId)
    
    {
        String betStr = betNumStr;
        if (betNumStr.contains(",")) {
            String[] list = betNumStr.split(",");
            StringBuffer numStrNum = new StringBuffer();
            for (String str : list) {
                long length = str.length();
                if (length % 2 != 0) {
                    return null;
                }
                
                for (int i = 0; i < (length / 2); i++) {
                    String str2 = str.substring(i * 2, i * 2 + 2);
                    if (i != (length / 2 - 1)) {
                        str2 = String.format("%s", str2);
                    }
                    numStrNum.append(str2);
                }
                numStrNum.append(",");
            }
            betStr = numStrNum.substring(0, numStrNum.length() - 1);
            
        } else {
            StringBuffer numStrNum = new StringBuffer();
            for (int i = 0; i < (betNumStr.length() / 2); i++) {
                String str2 = betNumStr.substring(i * 2, i * 2 + 2);
                if (i != (betNumStr.length() / 2 - 1)) {
                    str2 = String.format("%s ", str2);
                }
                numStrNum.append(str2);
            }
            betStr = numStrNum.toString();
        }
        return betStr;
    }
    
    public static String isPassZuXuan(String betStr, int playId)
    
    {
        switch (playId) {
            //单式
            case 2002:
            case 2006:
            case 2019:
            case 2020:
            case 2021:
            case 2022:
            case 2023:
            case 2024:
            case 2025:
            case 2026:
            
            {
                return betStr;
            }
            //组选   任选
            
            default: {
                String reStr = isZuXuanNum(betStr, true);
                return reStr;
            }
        }
    }
    
    public static boolean isPassPaozi(String betStr, long playId)
    
    {
        boolean isAllSame = isAllSameNum2(betStr);
        if (isAllSame) {
            return false;
        }
        return true;
    }
    
    
}
